tasks/main.yml

```diff
@@ -103,13 +103,15 @@
   command: "/opt/puppetlabs/bin/puppet cert list -a"
   when: not puppet_configured.stat.exists
 
-- name: run masterless puppet to configure the puppetmaster
-  command: "/opt/puppetlabs/bin/puppet apply --environment {{puppet_environment}} --pluginsync --hiera_config {{puppet_etc_dir}}/hiera.yaml --modulepath {{puppet_code_dir}}/environments/{{puppet_environment}}/modules {{puppet_code_dir}}/environments/{{puppet_environment}}/manifests"
+- name: run masterless puppet to configure the puppetmaster - NOTE! this will always create a change due to CCCP-2840
+  command: "/opt/puppetlabs/bin/puppet apply --detailed-exitcodes --environment {{puppet_environment}} --pluginsync --hiera_config {{puppet_etc_dir}}/hiera.yaml --modulepath {{puppet_code_dir}}/environments/{{puppet_environment}}/modules {{puppet_code_dir}}/environments/{{puppet_environment}}/manifests"
   register: puppet_result
   failed_when: (puppet_result.rc != 2) and (puppet_result.rc != 0)
+  changed_when: (puppet_result.rc >= 2)
```

Comment:
Based on https://puppet.com/docs/puppet/latest/man/agent.html
I'm not sure that all exit codes higher than 1 should be included in `changed when`.
In my mind exit code 2 is obvious that it should be a `changed_when` , 4 and 6 I think should be a failure.

For example if we get an exit code 4 in my mind that should be investigated.

I think the `changed_when` should be an `rc == 2 or rc == 4 or rc == 6` instead of `rc >= 2` since that communicate to the next reader that the writer have consider these exit states, who knows if puppet will at some point create a new exit code.

Reply:
Makes sense. Expanded accordingly.

It's really weird that they don't conform to the standard UNIX return code norm but, c'est la vie.

Reply:
> For example if we get an exit code 4 in my mind that should be investigated.

And for this, absolutely, this is covered by the current `failed_when` already.

Reply:
I don't know enough about standard UNIX return code norms but to me everything except `2` seems to fall into correct usage according to https://www.tldp.org/LDP/abs/html/exitcodes.html .